library/core/src/slice/mod.rs

```diff
@@ -542,10 +542,9 @@ impl<T> [T] {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
     pub fn swap(&mut self, a: usize, b: usize) {
-        // Can't take two mutable loans from one vector, so instead just cast
-        // them to their raw pointers to do the swap.
-        let pa: *mut T = &mut self[a];
-        let pb: *mut T = &mut self[b];
+        // Can't take two mutable loans from one vector, so instead use raw pointers.
+        let pa = ptr::raw_mut!(self[a]);
+        let pb = ptr::raw_mut!(self[b]);
```

Comment:
I'm just curious - the reason that this works, is that we use intrinsic indexing of the slice? If we would go through `IndexMut`, the problem wouldn't be solved (?)

Reply:
Yes, indexing a slice directly translates to a MIR instruction without any function calls. This is the same as what we do for arithmetic operators for primitive integers. The trait impls only exist for the type system and generics.